File: cpp/2_add_two_numbers/add_two_numbers.cpp
/*
 * @Date: 2023-07-02
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-07-02
 * @FilePath: /algorithm/cpp/2_add_two_numbers/add_two_numbers.cpp
 */

//  Definition for singly-linked list.
#include <cassert>
#include <tuple>
#include <vector>

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    ListNode *head = nullptr, *tail = nullptr;
    int carry = 0;
    while (l1 || l2) {
      int n1 = l1 ? l1->val : 0;
      int n2 = l2 ? l2->val : 0;
      int sum = n1 + n2 + carry;
      if (!head) {
        head = tail = new ListNode(sum % 10);
      } else {
        tail->next = new ListNode(sum % 10);
        tail = tail->next;
      }
      carry = sum / 10;
      if (l1) {
        l1 = l1->next;
      }
      if (l2) {
        l2 = l2->next;
      }
    }
    if (carry > 0) {
      tail->next = new ListNode(carry);
    }
    return head;
  }
};

bool is_same(ListNode* l1, ListNode* l2) {
  if (l1 == nullptr && l2 == nullptr) {
    return true;
  }
  if (l1 != nullptr && l2 != nullptr) {
    return l1->val == l2->val && is_same(l1->next, l2->next);
  }
  return false;
}

int main() {
  std::vector<std::tuple<ListNode*, ListNode*, ListNode*>> test_cases{
      {
          new ListNode(2, new ListNode(4, new ListNode(3))),
          new ListNode(5, new ListNode(6, new ListNode(4))),
          new ListNode(7, new ListNode(0, new ListNode(8))),
      },
      {
          new ListNode(0),
          new ListNode(0),
          new ListNode(0),
      },
      {
          new ListNode(
              9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9))))))),
          new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9)))),
          new ListNode(
              8, new ListNode(
                     9, new ListNode(
                            9, new ListNode(9, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(1)))))))),
      },
  };

  for (auto& [l1, l2, ans] : test_cases) {
    assert(is_same(Solution().addTwoNumbers(l1, l2), ans));
  }
}

File: golang/2_add_two_numbers/add_two_numbers.go
/*
 * @Date: 2023-07-02
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-07-02
 * @FilePath: /algorithm/golang/2_add_two_numbers/add_two_numbers.go
 */

// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1, l2 *ListNode) (head *ListNode) {
	var tail *ListNode
	carry := 0
	for l1 != nil || l2 != nil {
		n1, n2 := 0, 0
		if l1 != nil {
			n1 = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			n2 = l2.Val
			l2 = l2.Next
		}
		sum := n1 + n2 + carry
		sum, carry = sum%10, sum/10
		if head == nil {
			head = &ListNode{Val: sum}
			tail = head
		} else {
			tail.Next = &ListNode{Val: sum}
			tail = tail.Next
		}
	}
	if carry > 0 {
		tail.Next = &ListNode{Val: carry}
	}
	return
}

func isSame(l1, l2 *ListNode) bool {
	if l1 == nil && l2 == nil {
		return true
	}
	if l1 != nil && l2 != nil {
		return l1.Val == l2.Val && isSame(l1.Next, l2.Next)
	}
	return false
}

func main() {
	testMap := []struct {
		l1  *ListNode
		l2  *ListNode
		ans *ListNode
	}{
		{
			&ListNode{2, &ListNode{4, &ListNode{3, nil}}},
			&ListNode{5, &ListNode{6, &ListNode{4, nil}}},
			&ListNode{7, &ListNode{0, &ListNode{8, nil}}},
		},
		{
			&ListNode{0, nil},
			&ListNode{0, nil},
			&ListNode{0, nil},
		},
		{
			&ListNode{9, &ListNode{9, &ListNode{9, &ListNode{9, &ListNode{9, &ListNode{9, &ListNode{9, nil}}}}}}},
			&ListNode{9, &ListNode{9, &ListNode{9, &ListNode{9, nil}}}},
			&ListNode{8, &ListNode{9, &ListNode{9, &ListNode{9, &ListNode{0, &ListNode{0, &ListNode{0, &ListNode{1, nil}}}}}}}},
		},
	}

	for _, item := range testMap {
		assert.True(&testing.T{}, isSame(addTwoNumbers(item.l1, item.l2), item.ans))
	}
}
